File: solutions/0238-product-of-array-except-self/solution.py
from typing import List
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        izq= [1]*n
        der = [1]*n
        ans = [0]* n
        for i in range(1, n):
            izq[i] = izq[i-1] * nums[i-1]
        for i in range(n-2, -1, -1):
            der[i] = der[i+1] * nums[i+1]
        for i in range(0, n):
            ans[i] = izq[i] * der[i]

        
        return ans
    
class Solution2:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [1] * n
        for i in range(1, n):
            ans[i] = ans[i-1] * nums[i-1]
        der = 1
        for i in range(n-1, -1, -1):
            ans[i] *= der
            der *= nums[i]
        return ans

if __name__ == "__main__":
    s = Solution()

    print(s.productExceptSelf([1,2,3,4]))

File: solutions/0238-product-of-array-except-self/Solution.java
import java.util.*;

public class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        ans[0] = 1;
        for (int i = 1; i < n; i++)
            ans[i] = ans[i - 1] * nums[i - 1];
        int der = 1;
        for (int i = n - 1; i >= 0; i--) {
            ans[i] *= der;
            der *= nums[i];
        }
        return ans;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(Arrays.toString(
            sol.productExceptSelf(new int[]{1, 2, 3, 4})));
    }
}
